edward/criticisms/evaluate.py

```diff
@@ -115,26 +116,43 @@ def evaluate(metrics, data, n_samples=500, output_key=None):
   if metrics != ['log_lik'] and metrics != ['log_likelihood']:
     binary_discrete = (Bernoulli, Binomial)
     categorical_discrete = (Categorical, Multinomial, OneHotCategorical)
+    total_count = sess.run(getattr(output_key, 'total_count', tf.constant(1.)))
     if isinstance(output_key, binary_discrete + categorical_discrete):
       # Average over realizations of their probabilities, then predict
       # via argmax over probabilities.
       probs = [sess.run(output_key.probs, feed_dict) for _ in range(n_samples)]
-      probs = tf.add_n(probs) / tf.cast(n_samples, probs[0].dtype)
+      probs = np.sum(probs, axis=0) / n_samples
       if isinstance(output_key, binary_discrete):
         # make random prediction whenever probs is exactly 0.5
         random = tf.random_uniform(shape=tf.shape(probs))
         y_pred = tf.round(tf.where(tf.equal(0.5, probs), random, probs))
       else:
-        y_pred = tf.argmax(probs, len(probs.shape) - 1)
+        if total_count > 1:
+          mode = compute_multinomial_mode(probs, total_count, seed)
+          if len(output_key.sample_shape):
+            y_pred = tf.reshape(tf.tile(mode, output_key.sample_shape),
+                                [-1, len(probs)])
+          else:
+            y_pred = mode
+        else:
+          y_pred = tf.argmax(probs, len(probs.shape) - 1)
+      probs = tf.constant(probs)
     else:
       # Monte Carlo estimate the mean of the posterior predictive.
       y_pred = [sess.run(output_key, feed_dict) for _ in range(n_samples)]
       y_pred = tf.cast(tf.add_n(y_pred), y_pred[0].dtype) / \
           tf.cast(n_samples, y_pred[0].dtype)
+    if len(y_true.shape) == 0:
+      y_true = tf.expand_dims(y_true, 0)
+      y_pred = tf.expand_dims(y_pred, 0)
 
   # Evaluate y_true (according to y_pred if supervised) for all metrics.
   evaluations = []
   for metric in metrics:
+    if isinstance(metric, tuple):
```

Comment:
I personally don't think this API is intuitive and would default to not having the option exposed until we finalize ideas. That said, I don't feel too strongly about it so feel free to leave it here. However, do update the docstrings to reflect how to use this, and perhaps include an example in the docstrings' section.

Reply:
Done.